Firestore/core/test/firebase/firestore/util/path_test.cc

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+#define EXPECT_BASENAME_EQ(expected, source)                  \
+  do {                                                        \
+    EXPECT_EQ(std::string{expected}, Path::Basename(source)); \
+  } while (0)
+
+TEST(Path, Basename_NoSeparator) {
+  // POSIX would require all of these to be ".".
+  // python and libc++ agree this is "".
+  EXPECT_BASENAME_EQ("", "");
+  EXPECT_BASENAME_EQ("a", "a");
+  EXPECT_BASENAME_EQ("foo", "foo");
+  EXPECT_BASENAME_EQ(".", ".");
+  EXPECT_BASENAME_EQ("..", "..");
+}
+
+TEST(Path, Basename_LeadingSlash) {
+  EXPECT_BASENAME_EQ("", "/");
+  EXPECT_BASENAME_EQ("", "///");
+  EXPECT_BASENAME_EQ("a", "/a");
+  EXPECT_BASENAME_EQ("a", "//a");
+
+  EXPECT_BASENAME_EQ(".", "/.");
+  EXPECT_BASENAME_EQ("..", "/..");
+  EXPECT_BASENAME_EQ("..", "//..");
+}
+
+TEST(Path, Basename_IntermediateSlash) {
+  EXPECT_BASENAME_EQ("b", "/a/b");
+  EXPECT_BASENAME_EQ("b", "/a//b");
+  EXPECT_BASENAME_EQ("b", "//a/b");
+  EXPECT_BASENAME_EQ("b", "//a//b");
+
+  EXPECT_BASENAME_EQ("b", "//..//b");
+  EXPECT_BASENAME_EQ("b", "//a/./b");
+  EXPECT_BASENAME_EQ("b", "//a/.//b");
+}
+
+TEST(Path, Basename_TrailingSlash) {
+  // POSIX: "a/b//" => "b"
+  // python: "a/b//" => ""
+  // libc++: "a/b//" => "." (cppreference: "")
```

Comment:
The clear guidance from the aforementioned widely recognized standards is that each project should do its own thing, preferably incompatible with as many others as possible. I think you're leaning dangerously close to consistency with Python, why not do `//` instead?

More seriously, could you please clarify "cppreference" -- I presume you mean Filesystem TS as documented on cppreference, so perhaps specify that and also the function name?

Reply:
If you'd prefer I can make this wildly inconsistent with everything :-).

I've added text at the top of the test to clarify what I mean by the different cases.